oss-internship-2020/curl/tests/test_utils.cc

```diff
@@ -0,0 +1,269 @@
+      int setopt_write_data,
+      api_->curl_easy_setopt_ptr(curl_.get(), CURLOPT_WRITEDATA,
+                                 chunk_.PtrBoth()));
+  ASSERT_EQ(setopt_write_data, CURLE_OK);
+}
+
+void CurlTestUtils::CurlTestTearDown() {
+  // Cleanup curl
+  ASSERT_THAT(api_->curl_easy_cleanup(curl_.get()), sapi::IsOk());
+}
+
+void CurlTestUtils::PerformRequest(std::string& response) {
+  // Perform the request
+  SAPI_ASSERT_OK_AND_ASSIGN(int perform_code,
+                            api_->curl_easy_perform(curl_.get()));
+  ASSERT_EQ(perform_code, CURLE_OK);
+
+  // Get pointer to the memory chunk
+  sapi::v::GenericPtr remote_ptr;
+  remote_ptr.SetRemote(&((MemoryStruct*)chunk_.GetRemote())->memory);
+  ASSERT_THAT(sandbox_->TransferFromSandboxee(&remote_ptr), sapi::IsOk());
+  void* chunk_ptr = (void*)remote_ptr.GetValue();
+
+  // Get the string and store it in response
+  SAPI_ASSERT_OK_AND_ASSIGN(
+      response, sandbox_->GetCString(sapi::v::RemotePtr(chunk_ptr)));
+}
+
+void CurlTestUtils::PerformRequest() {
+  // If the response is not needed, pass a string that will be discarded
+  std::string discarded_response;
+  PerformRequest(discarded_response);
+}
+
+void CurlTestUtils::StartMockServer() {
+  // Get ai_list, a list of addrinfo structures, the port will be set later
+  addrinfo hints{AI_PASSIVE, AF_INET, SOCK_STREAM};
+  addrinfo* ai_list;
+  if (getaddrinfo("127.0.0.1", NULL, &hints, &ai_list) < 0) {
+    return;
+  }
+
+  // Loop over ai_list, until a socket is created
+  int listening_socket;
+  for (addrinfo* p = ai_list;; p = p->ai_next) {
+    if (p == nullptr) {
+      return;
+    }
+
+    // Try creating a socket
+    listening_socket = socket(p->ai_family, p->ai_socktype, p->ai_protocol);
+    if (listening_socket >= 0) {
+      // Try binding the socket to the address
+      if (bind(listening_socket, p->ai_addr, p->ai_addrlen) >= 0) {
+        // Assign an arbitrary available port to the socket address object
+        if (getsockname(listening_socket, p->ai_addr, &p->ai_addrlen) == -1) {
+          return;
+        }
+        port_ = ntohs(((struct sockaddr_in*)p->ai_addr)->sin_port);
+
+        break;
+
+      } else {
+        close(listening_socket);
+      }
+    }
+  }
+
+  freeaddrinfo(ai_list);
+
+  // Listen on the socket
+  if (listen(listening_socket, 256) == -1) {
+    return;
+  }
+
+  // Set server_thread_ operation to socket listening
+  server_thread_ = std::thread([=] {
+    // Create the master fd_set containing listening_socket
+    fd_set master_fd_set;
+    FD_ZERO(&master_fd_set);
+    FD_SET(listening_socket, &master_fd_set);
+
+    // Create an empty fd_set, will be used for making copies of master_fd_set
+    fd_set copy_fd_set;
+    FD_ZERO(&copy_fd_set);
+
+    int max_fd = listening_socket;
+
+    // Keep calling select and block after a new event happens
+    // Doesn't stop until the process doing tests is terminated
+    for (;;) {
+      copy_fd_set = master_fd_set;
```

Comment:
this is not defined per spec.
If the implementation used pointers inside fd_set, it would not work.